### PR TITLE
backend: reject colliding custom context names

### LINKED ISSUE #6427: backend: ContextStore silently overwrites colliding effective cluster names
## Describe the bug
`ContextStore` can silently overwrite one cluster with another when two contexts resolve to the same effective name, especially through `headlamp_info.customName`.

## To Reproduce
Steps to reproduce the bug:
1. Create or import two kubeconfig contexts with different backing identities but the same effective Headlamp name.
2. For a direct reproduction, set the same `headlamp_info.customName` value on both contexts.
3. Load those contexts through Headlamp so they pass through `LoadAndStoreKubeConfigs` and `ContextStore.AddContext`.
4. Query the stored contexts through the normal cluster listing flow.
5. Observe that only the last matching context remains visible because the earlier entry was overwritten.

## Environment (please provide info about your environment):
 - Installation type: Any installation that loads multiple kubeconfig contexts
 - Headlamp Version: Current `main` branch as of July 8, 2026
 - Other: Most visible in multi-cluster setups or imported kubeconfigs with duplicate `headlamp_info.customName` values

## Are you able to fix this issue?
Yes (I will propose a PR).

## Summary
`ContextStore.AddContext` derives the storage key from the effective cluster name and then unconditionally writes that key into the cache. If two different contexts resolve to the same effective name, the later insert silently replaces the earlier one instead of returning a validation error.

## Affected Area
- `backend/pkg/kubeconfig/contextStore.go`
- `backend/pkg/kubeconfig/kubeconfig.go`
- Cluster loading/import flows that call `ContextStore.AddContext`

## Expected Behaviour
Duplicate effective cluster names should be rejected deterministically so Headlamp does not hide one context behind another.

## Actual Behaviour
The later context silently overwrites the earlier one in the store, and subsequent reads only expose the last inserted context.

## Technical Evidence
- `backend/pkg/kubeconfig/contextStore.go` derives `name` from `headlamp_info.customName` and writes it with `cache.Set` without a duplicate guard.
- `backend/pkg/kubeconfig/kubeconfig.go` bulk-loads contexts via `LoadAndStoreKubeConfigs`, so imported kubeconfigs can hit this path immediately.
- `backend/pkg/kubeconfig/kubeconfig_test.go` already documents that duplicate kubeconfig context names can both be loaded before they reach the store.
- `gh issue view 2236` shows an older user-facing duplicate-context report, but this overwrite path still exists for effective-name collisions and is not tracked by an open issue.

## Suspected Root Cause
`ContextStore` treats the effective display name as a unique storage identifier but does not verify whether that key is already occupied by a different logical context.

## Proposed Direction
Reject inserts when another logical context already occupies the same effective name, while still allowing re-insertion of the same context during normal refresh/update flows.

## Acceptance Criteria
- Adding two different contexts with the same effective name returns a deterministic error instead of overwriting stored state.
- Re-adding the same logical context remains supported.
- Backend tests cover duplicate `headlamp_info.customName` collisions.

### PR DESCRIPTION
**Pull request by rootp1**
## Summary
- reject `ContextStore` collisions when different contexts resolve to the same effective context name
- keep re-adding the same logical context working for rename and refresh flows
- add backend regression coverage for custom-name collisions, non-custom collisions, and same-context updates

## Linked Issue
Fixes #6427

## What Changed
- added an effective-name helper in `backend/pkg/kubeconfig/contextStore.go`
- reject colliding inserts for duplicate effective names, including plain `Context.Name` collisions
- compare logical context identity so safe updates still overwrite the same stored context
- preserve existing dynamic manual replacement and dynamic kubeconfig reload flows
- added focused `contextStore` tests for collision rejection and same-context updates

## Verification
- `cd backend && go test ./pkg/kubeconfig/... -count=1` — passed
- `tmp=$(mktemp -d) && gomod=$(go env GOMODCACHE) && gocache=$(go env GOCACHE) && cd backend && HOME="$tmp" XDG_CONFIG_HOME="$tmp/.config" GOMODCACHE="$gomod" GOCACHE="$gocache" HEADLAMP_RUN_INTEGRATION_TESTS=true go test ./cmd -run 'TestDynamicClusters/override|TestRenameCluster' -count=1` — passed
- `cache=$(mktemp -d) && cd backend && GOLANGCI_LINT_CACHE="$cache" ./tools/golangci-lint run ./pkg/kubeconfig/...` — passed
- `cache=$(mktemp -d) && cd backend && GOLANGCI_LINT_CACHE="$cache" ./tools/golangci-lint run` — passed
- `gomod=$(go env GOMODCACHE) && gocache=$(go env GOCACHE) && tmp=$(mktemp -d) && cd backend && HOME="$tmp" XDG_CONFIG_HOME="$tmp/.config" GOMODCACHE="$gomod" GOCACHE="$gocache" HEADLAMP_RUN_INTEGRATION_TESTS=true go test ./...` — not a valid local full-suite analogue with isolated home; `backend/cmd` passed, then helm and portforward tests failed because the isolated home lacks the normal kubeconfig test setup

## Scope
- Only `backend/pkg/kubeconfig/contextStore.go` and `backend/pkg/kubeconfig/contextStore_test.go` were changed.

### CODE REVIEW

**Review by skoeva — CHANGES_REQUESTED** [review]
hi, could you take a look at the copilot comment and address the linting error in the CI?

File: backend/pkg/kubeconfig/contextStore.go [hunk 1]
```diff
@@ -78,41 +79,109 @@ func (c *contextStore) notifyListeners() {
...
-		}
+	name, usesCustomName, err := effectiveContextName(headlampContext)
+	if err != nil {
+		return err
+	}
 
-		// Now, decode the byte slice into your desired struct
-		var customObj CustomObject
+	existingContext, err := c.cache.Get(context.Background(), name)
+	existingName, existingUsesCustomName, existingNameErr := effectiveContextName(existingContext)
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go]: Thanks for pointing this out. I addressed it by serializing the context store mutators with the store mutex so the effective name check and the cache write now happen under the same lock. I also added a concurrent regression test that runs two same name inserts at once and verifies only one succeeds.

**Comment by rootp1** [comment]
@skoeva Thanks for the review. I addressed the Copilot comment by serializing AddContext and the other store mutators so the duplicate effective name check cannot race with the write. I also added a concurrent regression test for that path. The failing backend CI job was the wsl_v5 linting complaint in contextStore.go, and that is fixed in the latest push. I reran go test -v ./pkg/kubeconfig and ./tools/golangci-lint run ./pkg/kubeconfig/... locally on this branch.

File: backend/pkg/kubeconfig/contextStore.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go]: serialized AddContext and the other store mutators so the duplicate effective name check cannot race with the write.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


it looks like there's a merge-main commit in this PR — could you rebase onto main instead?

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

**Comment by rootp1** [comment]
> Thanks for working on this.
> 
> Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.

hi @illume, all the comments have been resolved, would you still want to gt this pr rebased instead of merged?

**Comment by rootp1** [comment]
@illume Thanks for the review. I have completed everything you pointed out, rebased the branch to remove the merge-main commit, and reran the kubeconfig tests locally. Please get back whenever you get a chance.

File: backend/pkg/kubeconfig/contextStore.go [hunk 2]
```diff
@@ -78,41 +79,124 @@ func (c *contextStore) notifyListeners() {
 
 // AddContext adds a context to the store.
 func (c *contextStore) AddContext(headlampContext *Context) error {
-	name := headlampContext.Name
+	name, usesCustomName, err := effectiveContextName(headlampContext)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:90]: Handled in the latest push. AddContext now rejects nil input up front with a deterministic error instead of dereferencing it later, and I added a regression test that covers AddContext(nil).

**Comment by rootp1** [comment]
@illume I picked up the remaining unattended Copilot comment on this PR and pushed a follow up. AddContext now returns a deterministic error for nil input instead of panicking, and there is a regression test covering that path.

**Comment by rootp1** [comment]
@illume CI is green on the current head and the earlier review items have been addressed. Could you please take another look when you have time?

**Review by rootp1 — COMMENTED** [review]
Addressed the nil-input case in `AddContextWithKeyAndTTL`.

Changes pushed in `5a7b64800`:
- return `errors.New("context cannot be nil")` before touching `headlampContext.Name`
- add a regression test covering `AddContextWithKeyAndTTL(nil, ...)`

This makes the TTL code path consistent with `AddContext` and avoids a nil-pointer panic on invalid input.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


A few of the commits don't quite follow the project guidelines. We use Linux kernel style for git commits — have a look at the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and previous commits with `git log`.

<details>
<summary>Commits that need attention</summary>

- `backend: guard nil TTL context input` — Description must start with a capital letter — e.g. `frontend: HomeButton: Fix the button` not `frontend: HomeButton: fix the button`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

File: backend/pkg/kubeconfig/contextStore.go [hunk 3]
```diff
@@ -170,15 +265,24 @@ func (c *contextStore) AddContextWithKeyAndTTL(headlampContext *Context, key str
 	// Keep the stored context identifier consistent with the cache key.
 	headlampContext.Name = key
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:363]: Thanks for flagging this. The current head already returns a deterministic error from `AddContextWithKeyAndTTL` on nil input and includes regression coverage for that path, so I resolved the stale thread to match the branch state. `cd backend && go test ./pkg/kubeconfig/...` passes on the current branch. @illume please have a look whenever you have a chance.

File: backend/pkg/kubeconfig/contextStore.go [hunk 4]
```diff
@@ -78,41 +80,128 @@ func (c *contextStore) notifyListeners() {
...
+	}
+
+	if customObj.CustomName != "" {
+		name = customObj.CustomName
+
+		return name, true, nil
+	}
+
+	return name, false, nil
+}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:163]: Thanks for pointing this out. I added a direct `*CustomObject` fast path in `effectiveContextName` and kept the JSON fallback for unstructured extension values, which removes the extra round trip from the common custom-name path. `cd backend && go test ./pkg/kubeconfig/...` passes on the updated commit. @illume please have a look whenever you have a chance.

**Comment by rootp1** [comment]
I picked up the backend CI failure on this branch. It was the gocritic single-case switch warning in contextStore, so I rewrote that fast path as an if type assertion, reran npm run backend:lint and go test ./pkg/kubeconfig/... locally, and both passed before pushing. @illume please have a look whenever you have a chance.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


The backend test job in CI is failing. Run `cd backend && go test ./...` to reproduce the errors locally.

<details>
<summary>How to run the backend tests</summary>

Run `cd backend && go test ./...` to see all failures. Fix the failing tests and commit the result.

</details>

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


Could you take a look at the commit messages in this PR? We follow a Linux kernel style for git commits — see the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and `git log` for examples.

<details>
<summary>Commits that need attention</summary>

- `backend: Rewrite context store type assertion` — Only one file changed inside `backend/`; add a sub-area so it's clear what was touched (e.g. `backend: ComponentName: description`).

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Comment by rootp1** [comment]
I cleaned this up into a single rebased commit with the title `backend: kubeconfig: Reject colliding context names` and force-pushed the branch. I also reran `npm run backend:lint` and `cd backend && go test ./...`, and both passed on the current head. @illume please have a look whenever you have a chance.

File: backend/pkg/kubeconfig/contextStore.go [hunk 5]
```diff
@@ -78,41 +80,136 @@ func (c *contextStore) notifyListeners() {
...
+	}
+
+	return fmt.Sprintf(
+		"fallback:%d|%s|%s|%s|%s",
+		headlampContext.Source,
+		headlampContext.KubeConfigPath,
+		clusterServer,
+		kubeCluster,
+		kubeAuthInfo,
+	)
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:295]: Thanks for flagging this. I updated the fallback logical identity to include `KubeContext.Namespace`, so contexts without a `ClusterID` no longer collapse together when they differ only by namespace. I reran `cd backend && go test ./pkg/kubeconfig/...`, `npm run backend:lint`, and `cd backend && go test ./...` on `81e18a165`, and all three passed.

File: backend/pkg/kubeconfig/contextStore_test.go [hunk 6]
```diff
@@ -167,3 +168,211 @@ func TestAddContextWithHeadlampInfo(t *testing.T) {
 	require.Error(t, err)
 	require.Equal(t, cache.ErrNotFound, err)
 }
+
+func TestAddContextRejectsDuplicateEffectiveName(t *testing.T) {
+	store := kubeconfig.NewContextStore()
+
+	customName := "shared-custom-name"
+	firstInfo := kubeconfig.CustomObject{CustomName: customName}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore_test.go:176]: Added that regression case in `81e18a165`. The new test covers two custom-named contexts with empty `ClusterID` values that differ only by namespace and verifies the collision is still rejected. `cd backend && go test ./pkg/kubeconfig/...` passes with the new coverage.

**Comment by rootp1** [comment]
@illume all the checks including the review from copilot is green, can you please take a look whenever you get a chance to

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for these changes.


Can you please address the open review comments? Once you've resolved each one, please mark it as resolved.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


A few of the commits don't quite follow the project guidelines. We use Linux kernel style for git commits — have a look at the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and previous commits with `git log`.

<details>
<summary>Commits that need attention</summary>

- `backend: reject all context name collisions` — Description must start with a capital letter — e.g. `frontend: HomeButton: Fix the button` not `frontend: HomeButton: fix the button`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

File: backend/pkg/kubeconfig/contextStore.go [hunk 7]
```diff
@@ -78,41 +80,139 @@ func (c *contextStore) notifyListeners() {
...
+	if existingNameErr != nil {
+		c.mu.Unlock()
+
+		return existingNameErr
+	}
+
+	if err == nil &&
+		(existingUsesCustomName || usesCustomName) &&
+		existingName == name &&
+		!isSameLogicalContext(existingContext, headlampContext) {
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go]: Handled in the current branch. AddContext now rejects duplicate effective names broadly, including plain Context.Name collisions, while keeping existing refresh paths working through logical identity checks.

File: backend/pkg/kubeconfig/contextStore.go [hunk 8]
```diff
@@ -78,41 +80,136 @@ func (c *contextStore) notifyListeners() {
...
 
-		// If the custom name is set, use it as the context name
-		if customObj.CustomName != "" {
-			name = customObj.CustomName
+	_, _, existingNameErr := effectiveContextName(existingContext)
+	if existingNameErr != nil {
+		c.mu.Unlock()
+
+		return existingNameErr
+	}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go]: Fixed in 73aadb79b. AddContext no longer reparses the stored context after the cache lookup, so the duplicate decision relies on the cache key and logical identity check.

File: backend/pkg/kubeconfig/contextStore.go [hunk 9]
```diff
@@ -78,41 +80,136 @@ func (c *contextStore) notifyListeners() {
...
-		info := headlampContext.KubeContext.Extensions["headlamp_info"]
-		// Convert the runtime.Unknown object to a byte slice
-		unknownBytes, err := json.Marshal(info)
-		if err != nil {
-			return err
-		}
+	name, _, err := effectiveContextName(headlampContext)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go]: The stricter behavior is intended. I updated the PR description to say duplicate effective names are rejected broadly, including plain Context.Name collisions.

**Comment by rootp1** [comment]
@illume CI and review comments are green now. Please take a look whenever you have a chance to.

**Comment by rootp1** [comment]
@illume I reworded the commit messages to match the contribution guidelines. The force push retriggered CI, so I will wait for the new checks to finish.

**Comment by rootp1** [comment]
@illume I checked the current branch commits and the commit titles on this PR are in the project format now.

**Comment by rootp1** [comment]
@illume @vyncent-t I squashed this PR history into a single clean commit and updated the commit title to match the contribution guidelines. The branch has been force pushed and is ready for another look.

**Comment by rootp1** [comment]
@illume just a quick reminder on this one when you have a chance. It has already been through several review rounds, I addressed the feedback, and I tested the current branch locally. If it looks good from your side now, could you please merge it?

**Comment by rootp1** [comment]
I noticed the head commit was missing a DCO Signed off by trailer, so I amended it in place and force pushed with the same content, tests still pass locally with go test ./pkg/kubeconfig/... on the backend. CI is running again on the new commit. @illume @sniok @ashu8912 @vyncent-t whenever you get a chance, could one of you take another look now that the earlier review rounds and commit formatting issues are addressed.

**Comment by rootp1** [comment]
@illume @vyncent-t this one's been through many review rounds now, all resolved, checks green. Would appreciate a decision on merge readiness when you have time.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


The open review comments from Copilot still need attention — can you have a look? Once addressed, please mark them as resolved.

File: backend/pkg/kubeconfig/contextStore.go [hunk 10]
```diff
@@ -78,41 +80,173 @@ func (c *contextStore) notifyListeners() {
...
+
+	return fmt.Sprintf(
+		"fallback:%d|%s|%s|%s|%s|%s",
+		headlampContext.Source,
+		headlampContext.KubeConfigPath,
+		clusterServer,
+		kubeCluster,
+		kubeAuthInfo,
+		kubeNamespace,
+	)
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:295]: Good catch, switched the fallback identity to render each field with %q so a literal pipe or quote inside a path or name can no longer collide two different identities.

File: backend/pkg/kubeconfig/contextStore.go [hunk 11]
```diff
@@ -78,41 +80,173 @@ func (c *contextStore) notifyListeners() {
...
+	return fallbackContextIdentity(existingContext) == fallbackContextIdentity(newContext)
+}
+
+func isDynamicManualUpdate(existingContext, newContext *Context, newContextName string) bool {
+	return existingContext.Source == DynamicCluster &&
+		newContext.Source == DynamicCluster &&
+		existingContext.ClusterID == "" &&
+		newContext.ClusterID == "" &&
+		existingContext.Name == newContextName
+}
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:209]: You are right that name plus empty ClusterID alone was not enough, so isDynamicManualUpdate now also requires the cluster server to match before treating it as an update rather than a collision.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


The open review comments from Copilot still need attention — can you have a look? Once addressed, please mark them as resolved.

**Comment by rootp1** [comment]
I pushed a follow up for the remaining Copilot issue. Manual dynamic cluster updates now keep their existing replace behavior, custom-name collisions are still rejected, and I reran npm run backend:lint, cd backend && go test ./..., and npm run backend:build locally before pushing.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


There are some open Copilot review comments — could you take a look at them? Please mark each one as resolved once you've addressed it.

File: backend/pkg/kubeconfig/contextStore.go [hunk 12]
```diff
@@ -78,41 +80,221 @@ func (c *contextStore) notifyListeners() {
...
 	}
 
+	if err != nil && err != cache.ErrNotFound {
+		c.mu.Unlock()
+
+		return err
+	}
+
 	// Keep the stored context identifier consistent with the cache key.
 	headlampContext.Name = name
```
[the remark this answers is not in the document]
Comment (rootp1) [reply backend/pkg/kubeconfig/contextStore.go:112]: This overwrite is pre-existing behavior in AddContext from before this PR, not something introduced by the collision checks added here. Fixing it properly means touching Context's originalName propagation through getClusters and the delete flow in backend/cmd/headlamp.go, which is outside the two files this PR touches. I'd rather keep this PR scoped to the collision rejection logic and handle the originalName issue separately.
